internal/sqladapter/collection.go

```diff
@@ -49,63 +82,199 @@ func (c *collection) PrimaryKeys() []string {
 	return c.pk
 }
 
+func (c *collection) filterConds(conds ...interface{}) []interface{} {
+	if tr, ok := c.PartialCollection.(condsFilter); ok {
+		return tr.FilterConds(conds...)
+	}
+	if len(conds) == 1 && len(c.pk) == 1 {
+		if id := conds[0]; IsKeyValue(id) {
+			conds[0] = db.Cond{c.pk[0]: id}
+		}
+	}
+	return conds
+}
+
 // Find creates a result set with the given conditions.
 func (c *collection) Find(conds ...interface{}) db.Result {
+	if c.err != nil {
+		res := &Result{}
+		res.setErr(c.err)
+		return res
+	}
 	return NewResult(
-		c.p.Database(),
-		c.p.Name(),
-		c.p.Conds(conds...),
+		c.Database(),
+		c.Name(),
+		c.filterConds(conds...),
 	)
 }
 
 // Exists returns true if the collection exists.
 func (c *collection) Exists() bool {
-	if err := c.p.Database().TableExists(c.p.Name()); err != nil {
+	if err := c.Database().TableExists(c.Name()); err != nil {
 		return false
 	}
 	return true
 }
 
 // InsertReturning inserts an item and updates the given variable reference.
 func (c *collection) InsertReturning(item interface{}) error {
-	if reflect.TypeOf(item).Kind() != reflect.Ptr {
-		return fmt.Errorf("Expecting a pointer to map or string but got %T", item)
+	if item == nil || reflect.TypeOf(item).Kind() != reflect.Ptr {
+		return fmt.Errorf("Expecting a pointer but got %T", item)
+	}
+
+	// Grab primary keys
+	pks := c.PrimaryKeys()
+	if len(pks) == 0 {
+		return fmt.Errorf("InsertReturning: Cannot update an item without primary keys")
 	}
 
 	var tx DatabaseTx
 	inTx := false
 
-	if currTx := c.p.Database().Transaction(); currTx != nil {
-		tx = newTxWrapper(c.p.Database())
+	if currTx := c.Database().Transaction(); currTx != nil {
+		tx = NewDatabaseTx(c.Database())
 		inTx = true
 	} else {
 		// Not within a transaction, let's create one.
 		var err error
-		tx, err = c.p.Database().NewLocalTransaction()
+		tx, err = c.Database().NewDatabaseTx(c.Database().Context())
 		if err != nil {
 			return err
 		}
 		defer tx.(Database).Close()
 	}
 
-	var res db.Result
+	// Allocate a clone of item.
+	newItem := reflect.New(reflect.ValueOf(item).Elem().Type()).Interface()
```

Comment:
`InsertReturning(&item)` updates `item` after inserting it by pulling a fresh `item` from the db, this populates auto values, like dates or IDs.

The problem is that this `item` gets reset when we pull it, so if `item` had a special non-db field, this field would get reset as well. For instance:

```
item = Post{
  Title: "foo",
  Children: []Post{ child1, child2, ... }
}

sess.InsertReturning(&item)
```

Let's suppose `Title` gets mapped to the `title` column, but `Children` isn't inserted, if we reset `item` we'd lose `Children`, in order not to do that we use a clone to pull whatever we expect from the database, and then we overwrite specific fields on the original `item`, this keeps `Children` intact.